File: apps/mobile/lib/auth/auth_client.dart
import 'package:built_value/serializer.dart';
import 'package:dio/dio.dart';
import 'package:omnistock_api_client/omnistock_api_client.dart';

import 'auth_exceptions.dart';
import 'token_store.dart';

/// T-001-17 ★ (L-3 — cold-start restore / refresh-failure UX). Distinguishes
/// a REAL dead session from a merely-transient failure so callers (cold-start
/// bootstrap, any silent-refresh call site) don't force a full re-login/wipe
/// over a flaky network — mirrors [AuthClient._doRefresh]'s existing storage
/// behavior (only a genuine `401 INVALID_REFRESH` wipes storage) by exposing
/// that distinction on the wire, not just internally.
enum RefreshOutcome {
  /// Refresh succeeded — a fresh access/refresh token pair is stored.
  success,

  /// The refresh token is genuinely dead (expired/revoked/reuse-detected, or
  /// simply absent) — storage has been wiped; the caller must treat the
  /// session as over (ux-wireframe §7: bounce to login + polite toast).
  sessionExpired,

  /// Network failure / 5xx / 429 — the refresh token was intentionally left
  /// untouched in storage (still potentially valid). The caller should offer
  /// a retry/offline state, NOT a forced re-login/wipe.
  transientFailure,
}

/// T-001-17 ★ — mobile token/refresh flow.
///
/// Contract: docs/features/F-001/api-spec.md §0 (LOCKED). Skill:
/// client-security. Mirrors the shape of `apps/web/src/lib/auth-client.ts`
/// (same retry-once/session-expired semantics) but the transport is BODY,
/// not cookie (D-019's cookie-path fix is web-only — see api-spec §0 item 2
/// + ux-wireframe §8): mobile always omits `tokenTransport` (server default
/// `"body"`) and stores the plaintext refresh token in the OS
/// keychain/keystore via [TokenStore], never in a cookie.
///
/// Consumes the GENERATED [AuthApi] client only — no hand-reshaping of
/// request/response types (frontend does not decide API/data shape).
class AuthClient {
  AuthClient({
    required AuthApi authApi,
    required TokenStore tokenStore,
    String? deviceId,
  })  : _authApi = authApi,
        _tokenStore = tokenStore,
        _deviceId = deviceId;

  final AuthApi _authApi;
  final TokenStore _tokenStore;
  final String? _deviceId;

  TokenStore get tokenStore => _tokenStore;

  /// T-001-17 ★ (L-4): the label THIS client instance sends as `deviceId` on
  /// login/refresh (api-spec §2.2/§2.3) — echoed back verbatim as
  /// `Session.deviceId` (api-spec §2.6). Since `GET /auth/sessions` marks
  /// `current` from the `omni_rt` cookie only (api-spec §2.6: "null if no
  /// cookie, e.g. a Bearer-only mobile call"), mobile's `current` is always
  /// `false` server-side — comparing a listed session's `deviceId` against
  /// THIS getter is the only client-side signal mobile has for "is this row
  /// the device I'm holding", and is exposed here so the session-list UI can
  /// use it instead of trusting the always-false `current` flag.
  String? get deviceId => _deviceId;

  int? _retryAfterFromHeaders(Map<String, List<String>> headers) {
    final values = headers['retry-after'] ?? headers['Retry-After'];
    if (values == null || values.isEmpty) return null;
    return int.tryParse(values.first);
  }

  ApiError _mapDioError(DioException e) {
    final response = e.response;
    if (response == null) {
      // Network failure / no response at all — no ErrorResponse to parse.
      return ApiError(0, null);
    }
    final status = response.statusCode ?? 0;
    final retryAfter = _retryAfterFromHeaders(response.headers.map);
    return ApiError(status, _parseErrorResponse(response.data), retryAfter);
  }

  /// The generated [AuthApi] methods only deserialize the SUCCESS-path body
  /// (built_value's typed round-trip) — on a non-2xx response Dio hands us
  /// the raw transformed JSON (`Map<String, dynamic>`, since
  /// `receiveDataWhenStatusError` defaults true), not an [ErrorResponse]
  /// instance. This mirrors the JSON parsing `apps/web/src/lib/auth-client.ts`
  /// does itself (`await res.json()`) for the same reason (its client
  /// doesn't build a typed error object either).
  ErrorResponse? _parseErrorResponse(Object? data) {
    if (data is ErrorResponse) return data;
    if (data is Map) {
      try {
        return standardSerializers.deserialize(
          data,
          specifiedType: const FullType(ErrorResponse),
        ) as ErrorResponse;
      } catch (_) {
        return null;
      }
    }
    return null;
  }

  // ---- Public endpoints (no access token needed) ----

  Future<SignupResponse> signup({
    required String email,
    required String password,
  }) async {
    try {
      final res = await _authApi.authSignup(
        signupRequest: SignupRequest((b) => b
          ..email = email
          ..password = password),
      );
      final body = res.data;
      if (body == null) {
        throw ApiError(res.statusCode ?? 0, null);
      }
      return body;
    } on DioException catch (e) {
      throw _mapDioError(e);
    }
  }

  /// US-2. Mobile ALWAYS uses body transport (omit `tokenTransport`, server
  /// default `"body"`, api-spec §0 item 2) — never declares `"cookie"`.
  /// On success, the access token goes to memory and the refresh token to
  /// the OS keychain/keystore (client-security skill).
  Future<TokenResponse> login({
    required String email,
    required String password,
  }) async {
    try {
      final res = await _authApi.authLogin(
        loginRequest: LoginRequest((b) {
          b
            ..email = email
            ..password = password
            ..tokenTransport = LoginRequestTokenTransportEnum.body;
          if (_deviceId != null) b.deviceId = _deviceId;
        }),
      );
      final body = res.data;
      if (body == null) {
        throw ApiError(res.statusCode ?? 0, null);
      }
      _tokenStore.setAccessToken(body.accessToken, expiresInSeconds: body.expiresIn);
      final refreshToken = body.refreshToken;
      if (refreshToken != null) {
        await _tokenStore.setRefreshToken(refreshToken);
      }
      return body;
    } on DioException catch (e) {
      throw _mapDioError(e);
    }
  }

  /// Rotates the refresh token. Presents the CURRENT refresh token from
  /// secure storage in the request body (body transport, api-spec §2.3).
  /// Returns true + stores the new pair on success; on failure, per mobile
  /// guidance M-2 (api-spec §5): **do not retry with the same token** — wipe
  /// storage. Never throws; callers treat it as a boolean gate (mirrors
  /// `silentRefresh` in apps/web/src/lib/auth-client.ts).
  ///
  /// This is a thin bool projection of [silentRefreshDetailed] kept for
  /// existing callers ([requestWithRefresh]) that only ever wipe/retry on a
  /// TRUE dead session anyway — a transient failure and a dead session both
  /// mean "this in-flight request can't be retried right now" from that call
  /// site's point of view. Callers that need to react differently to a
  /// transient failure (L-3 — cold-start bootstrap) should call
  /// [silentRefreshDetailed] directly instead.
  Future<bool> silentRefresh() async {
    final outcome = await silentRefreshDetailed();
    return outcome == RefreshOutcome.success;
  }

  /// Same single-flight refresh as [silentRefresh], but surfaces the
  /// [RefreshOutcome] distinction (L-3) instead of collapsing
  /// session-expired and transient-network-failure into the same `false`.
  Future<RefreshOutcome> silentRefreshDetailed() async {
    if (_inflightRefresh != null) return _inflightRefresh!;
    final future = _doRefresh();
    _inflightRefresh = future;
    try {
      return await future;
    } finally {
      _inflightRefresh = null;
    }
  }

  // Single-flight dedupe (mirrors apps/web/src/lib/auth-client.ts): several
  // concurrent 401s must share ONE /auth/refresh call, or a second/third
  // real refresh call would hit the just-rotated token as a benign-retry 401
  // (arch §3.5 leeway window) purely from a client-side race, wrongly
  // throwing SessionExpiredException for a still-live session.
  Future<RefreshOutcome>? _inflightRefresh;

  // T-001-17 ★ (L-2): monotonically-bumped logout epoch. `_doRefresh` reads
  // the epoch BEFORE awaiting the network call and re-checks it right before
  // persisting the rotated pair — if a logout (`_wipeForLogout`) intervened
  // during the await, the epoch will have moved and we must NOT write a
  // freshly-rotated token back into storage that was just deliberately
  // wiped (a refresh completing after `clearAll()` would otherwise
  // resurrect a "logged out" session).
  int _logoutEpoch = 0;

  /// Every code path that wipes storage because the user is logging out
  /// (as opposed to a dead/expired session being cleaned up) goes through
  /// here so the epoch bump and the wipe always happen together.
  Future<void> _wipeForLogout() async {
    _logoutEpoch++;
    await _tokenStore.clearAll();
  }

  Future<RefreshOutcome> _doRefresh() async {
    final epochAtStart = _logoutEpoch;
    final currentRefreshToken = await _tokenStore.getRefreshToken();
    if (currentRefreshToken == null) {
      await _tokenStore.clearAll();
      return RefreshOutcome.sessionExpired;
    }
    try {
      final res = await _authApi.authRefresh(
        refreshRequest: RefreshRequest((b) {
          b.refreshToken = currentRefreshToken;
          if (_deviceId != null) b.deviceId = _deviceId;
        }),
      );
      final body = res.data;
      if (body == null) {
        // D-022 ★ re-review fix (Minor #5): a 200 status with no/unparseable
        // body is a broken proxy/CDN symptom (malformed success), NOT the
        // server's documented dead-refresh-token signal (that's a real 401,
        // handled below). Treat it the same as network/5xx/429: leave
        // storage untouched and let the caller offer retry — wiping a
        // possibly-still-live session over a malformed 200 would be strictly
        // worse than a spurious retry prompt.
        return RefreshOutcome.transientFailure;
      }
      if (_logoutEpoch != epochAtStart) {
        // A logout happened while this refresh was in flight — storage was
        // already deliberately wiped; do not rewrite a dead token into it.
        return RefreshOutcome.sessionExpired;
      }
      _tokenStore.setAccessToken(body.accessToken, expiresInSeconds: body.expiresIn);
      final rotated = body.refreshToken;
      if (rotated != null) {
        await _tokenStore.setRefreshToken(rotated);
      }
      return RefreshOutcome.success;
    } on DioException catch (e) {
      // 401 INVALID_REFRESH (expired/revoked/reuse-detected — all
      // indistinguishable on the wire, api-spec §2.3) → the session is
      // unrecoverable client-side. Mobile guidance M-2: wipe, don't retry
      // with the same (now possibly-consumed) token.
      final status = e.response?.statusCode;
      if (status == 401) {
        if (_logoutEpoch == epochAtStart) {
          await _tokenStore.clearAll();
        }
        return RefreshOutcome.sessionExpired;
      }
      // Network/5xx/429 (L-3): leave storage untouched — a transient failure
      // should not destroy a still-potentially-valid refresh token. Most call
      // sites (requestWithRefresh, via the silentRefresh bool projection)
      // still surface this as a failed refresh for THIS in-flight request;
      // callers that can distinguish (cold-start bootstrap) get
      // transientFailure instead of sessionExpired so they can offer a
      // retry/offline state rather than forcing a full re-login.
      return RefreshOutcome.transientFailure;
    }
  }

  /// Runs [send] once; on a 401 [ApiError] that [isAuthExpiry] accepts,
  /// attempts exactly one silent refresh then retries [send] exactly once. A
  /// second failure — refresh itself failing, or the retried call still
  /// 401-with-[isAuthExpiry]-true — surfaces as [SessionExpiredException].
  /// Never loops (client-security skill).
  ///
  /// [isAuthExpiry] defaults to "every 401 means the access token died" —
  /// correct for endpoints whose ONLY 401 source is `JwtAuthGuard` rejecting
  /// a dead/missing Bearer token (`getSessions`, `logoutAll`). It is
  /// overridden for `changePassword` (api-spec §2.7), whose 401 is
  /// double-duty: the SAME status also means "wrong `currentPassword`" (an
  /// `INVALID_CREDENTIALS`-coded semantic error, nothing to do with the
  /// access token) — treating that as session-expiry would wipe a live
  /// session and kick the user to login over a simple typo.
  Future<T> requestWithRefresh<T>(
    Future<T> Function() send, {
    bool Function(ApiError) isAuthExpiry = _defaultIsAuthExpiry,
  }) async {
    try {
      return await send();
    } on ApiError catch (first) {
      if (first.status != 401 || !isAuthExpiry(first)) rethrow;

      final refreshed = await silentRefresh();
      if (!refreshed) {
        throw const SessionExpiredException();
      }

      try {
        return await send();
      } on ApiError catch (second) {
        if (second.status == 401 && isAuthExpiry(second)) {
          // Refreshed successfully but the retried call still 401'd (e.g.
          // kicked mid-flight, ux-wireframe §7) — do not loop again.
          await _tokenStore.clearAll();
          throw const SessionExpiredException();
        }
        rethrow;
      }
    }
  }

  static bool _defaultIsAuthExpiry(ApiError _) => true;

  /// `changePassword`'s 401 is only an access-token-expiry signal when it
  /// carries NO app-level error code (a bare `JwtAuthGuard` rejection body).
  /// A 401 that DOES carry `INVALID_CREDENTIALS` is the documented
  /// wrong-current-password outcome (api-spec §2.7) — never route that
  /// through silent-refresh/session-expiry.
  static bool _isChangePasswordAuthExpiry(ApiError e) => e.code != 'INVALID_CREDENTIALS';

  // ---- Authenticated endpoints (Bearer + silent-refresh-then-retry-once) ----

  Options _bearerOptions() {
    final token = _tokenStore.accessToken;
    return Options(headers: token != null ? {'Authorization': 'Bearer $token'} : null);
  }

  Future<SessionsResponse> getSessions() {
    return requestWithRefresh(() async {
      try {
        final res = await _authApi.authSessions(headers: _bearerOptions().headers);
        final body = res.data;
        if (body == null) throw ApiError(res.statusCode ?? 0, null);
        return body;
      } on DioException catch (e) {
        throw _mapDioError(e);
      }
    });
  }

  /// Per-device / current-device logout (US-4). Body-transport: presents
  /// the current refresh token so the server can resolve "current family"
  /// even without the Bearer token attached (mirrors api-spec §2.4 — the
  /// endpoint itself has `security: []`). Optional [familyId] targets a
  /// specific LISTED non-current session (§2.4/§2.6, M-3 ownership-checked
  /// server-side).
  Future<void> logoutDevice({String? familyId}) async {
    final currentRefreshToken = await _tokenStore.getRefreshToken();
    try {
      await _authApi.authLogout(
        logoutRequest: LogoutRequest((b) {
          if (currentRefreshToken != null) b.refreshToken = currentRefreshToken;
          if (familyId != null) b.familyId = familyId;
        }),
      );
    } on DioException catch (e) {
      throw _mapDioError(e);
    }
    // Only wipe local state when logging out the CURRENT device (no
    // familyId, or it targets the session we're on) — logging out a listed
    // OTHER device must not clear our own live session (ux-wireframe §4).
    if (familyId == null) {
      await _wipeForLogout();
    }
  }

  Future<void> logoutAll() async {
    await requestWithRefresh(() async {
      try {
        await _authApi.authLogoutAll(headers: _bearerOptions().headers);
        return null;
      } on DioException catch (e) {
        throw _mapDioError(e);
      }
    });
    await _wipeForLogout();
  }

  Future<OkResponse> changePassword({
    required String currentPassword,
    required String newPassword,
  }) {
    return requestWithRefresh(() async {
      final currentRefreshToken = await _tokenStore.getRefreshToken();
      try {
        final res = await _authApi.authChangePassword(
          changePasswordRequest: ChangePasswordRequest((b) {
            b
              ..currentPassword = currentPassword
              ..newPassword = newPassword;
            if (currentRefreshToken != null) b.refreshToken = currentRefreshToken;
          }),
          headers: _bearerOptions().headers,
        );
        final body = res.data;
        if (body == null) throw ApiError(res.statusCode ?? 0, null);
        return body;
      } on DioException catch (e) {
        throw _mapDioError(e);
      }
    }, isAuthExpiry: _isChangePasswordAuthExpiry);
  }
}

File: apps/web/src/lib/auth-client.ts
/**
 * T-001-16 ★ — web token/cookie refresh flow.
 *
 * Contract: docs/features/F-001/api-spec.md §0 (LOCKED). Skill: client-security.
 *
 *  - Access token: in-memory only (token-store.ts), sent as
 *    `Authorization: Bearer <token>`.
 *  - Refresh token: httpOnly `omni_rt` cookie (web always sends
 *    `tokenTransport: "cookie"` at login) — this module never reads/stores a
 *    refresh-token string; it relies on the browser sending the cookie
 *    automatically (`credentials: "include"`, same-origin via the proxy).
 *  - `X-CSRF-Token` header (double-submit) is attached on every
 *    cookie-borne mutation that carries the refresh token: refresh, logout,
 *    change-password. `logout-all` is Bearer-only (access token, no refresh
 *    token / cookie involved, api-spec §2.5) so it does NOT get the CSRF
 *    header — there is no cookie authority to forge there.
 *  - On a 401 from an authenticated call, we attempt exactly ONE silent
 *    refresh, then retry the original call ONCE. A second failure surfaces
 *    to the caller as a session-expired condition — never loop, never retry
 *    more than once (client-security skill). Concurrent 401s across multiple
 *    in-flight requests share a single `/auth/refresh` call (single-flight
 *    dedupe, client-security review Important #2) so a loser never wrongly
 *    hits the benign-retry 401 and gets kicked out of a live session.
 *
 * We deliberately do NOT reshape the generated contract types — every
 * request/response body typed here comes from `@omnistock/contracts`
 * (`components["schemas"][...]`).
 */
import type { components } from "@omnistock/contracts";
import { AUTH_BASE } from "./api-base";
import { csrfHeader } from "./csrf";
import { clearAccessToken, getAccessToken, setAccessToken } from "./token-store";

export type SignupRequest = components["schemas"]["SignupRequest"];
export type SignupResponse = components["schemas"]["SignupResponse"];
export type LoginRequest = components["schemas"]["LoginRequest"];
export type TokenResponse = components["schemas"]["TokenResponse"];
export type SessionsResponse = components["schemas"]["SessionsResponse"];
export type ChangePasswordRequest = components["schemas"]["ChangePasswordRequest"];
export type ErrorResponse = components["schemas"]["ErrorResponse"];
export type OkResponse = components["schemas"]["OkResponse"];

/** Thrown for any non-2xx response; carries the parsed error envelope (when
 * present) and the HTTP status + Retry-After (for 429 throttle UX). */
export class ApiError extends Error {
  status: number;
  code?: string;
  retryAfterSeconds?: number;

  constructor(status: number, body: ErrorResponse | null, retryAfterSeconds?: number) {
    super(body?.error.message ?? `Request failed with status ${status}`);
    this.name = "ApiError";
    this.status = status;
    this.code = body?.error.code;
    this.retryAfterSeconds = retryAfterSeconds;
  }
}

/** Thrown when silent refresh + retry-once both fail — the caller (screen /
 * router) should route to /login with the session-expired toast
 * (ux-wireframe §7). */
export class SessionExpiredError extends Error {
  constructor() {
    super("session expired");
    this.name = "SessionExpiredError";
  }
}

/**
 * Shared predicate so every screen handles a dead session identically
 * (ux-wireframe §7: "เด้งออกไปหน้า /login ทันที + toast สุภาพ ... ใช้ข้อความ
 * เดียวกันทุกสาเหตุ") — client-security review Important #4. Screens should
 * `catch` a failed auth-client call, check `isSessionExpired(err)` FIRST
 * (before any other error-shape handling), and if true: clear local state,
 * redirect to `/login`, and show `auth.sessionExpired.toast` — never retry
 * against a session that's already confirmed dead.
 */
export function isSessionExpired(err: unknown): err is SessionExpiredError {
  return err instanceof SessionExpiredError;
}

function parseRetryAfter(res: Response): number | undefined {
  const header = res.headers.get("Retry-After");
  if (!header) return undefined;
  const seconds = Number(header);
  return Number.isFinite(seconds) ? seconds : undefined;
}

async function parseErrorBody(res: Response): Promise<ErrorResponse | null> {
  try {
    return (await res.json()) as ErrorResponse;
  } catch {
    return null;
  }
}

interface RequestOptions {
  auth?: boolean; // attach Authorization: Bearer <access token>
  csrf?: boolean; // attach X-CSRF-Token (cookie-borne mutations)
}

/**
 * All 8 F-001 auth endpoints — every `path` this module calls with is one of
 * these, so `rawRequest` can unconditionally prefix with `AUTH_BASE`
 * (`/auth`) rather than `API_BASE` (`/api`). Kept as an explicit allowlist
 * (not just "everything starts with /auth") so a future non-auth call added
 * to this file by mistake fails loudly in review rather than silently
 * path-missing `omni_rt`'s `Path=/auth` scope again.
 */
const AUTH_PATHS = new Set([
  "/signup",
  "/login",
  "/refresh",
  "/logout",
  "/logout-all",
  "/sessions",
  "/change-password",
]);

async function rawRequest(
  path: string,
  init: RequestInit,
  opts: RequestOptions,
): Promise<Response> {
  const headers = new Headers(init.headers);
  headers.set("Content-Type", "application/json");
  if (opts.auth) {
    const token = getAccessToken();
    if (token) headers.set("Authorization", `Bearer ${token}`);
  }
  if (opts.csrf) {
    for (const [k, v] of Object.entries(csrfHeader())) headers.set(k, v);
  }
  if (!AUTH_PATHS.has(path)) {
    throw new Error(`auth-client: unrecognized path "${path}" — not in AUTH_PATHS allowlist`);
  }
  return fetch(`${AUTH_BASE}${path}`, {
    ...init,
    headers,
    credentials: "include", // send/receive omni_rt + omni_csrf cookies
  });
}

let inflightRefresh: Promise<boolean> | null = null;

async function doRefresh(): Promise<boolean> {
  const res = await rawRequest(
    "/refresh",
    { method: "POST", body: JSON.stringify({}) },
    { csrf: true },
  );
  if (!res.ok) {
    clearAccessToken();
    return false;
  }
  const body = (await res.json()) as TokenResponse;
  setAccessToken(body.accessToken, body.expiresIn);
  return true;
}

/**
 * Calls POST /auth/refresh. On success, stores the new access token and
 * returns true. On failure, clears the access token and returns false — it
 * never throws, so callers can treat it as a plain boolean gate.
 *
 * Single-flight (client-security review Important #2): concurrent callers
 * (e.g. several authenticated requests all 401-ing around the same moment)
 * share ONE in-flight `/auth/refresh` call instead of each firing their own
 * — a second/third real refresh call would hit the just-rotated token as a
 * benign-retry `401` (arch §3.5 leeway window) purely due to a client-side
 * race, which would wrongly throw `SessionExpiredError` for a session that
 * is actually still alive. The in-flight promise is cleared once settled so
 * the NEXT genuine 401 (a later moment in time) starts a fresh refresh.
 */
export async function silentRefresh(): Promise<boolean> {
  if (inflightRefresh) return inflightRefresh;
  inflightRefresh = doRefresh().finally(() => {
    inflightRefresh = null;
  });
  return inflightRefresh;
}

/**
 * Authenticated request with silent-refresh-then-retry-once (T-001-16,
 * ux-wireframe §7). Pure orchestration over an injectable `send` so the
 * retry/refresh decision tree is unit-testable without a real network (see
 * auth-client.retry.test.ts).
 */
export async function requestWithRefresh(
  send: () => Promise<Response>,
  refresh: () => Promise<boolean> = silentRefresh,
): Promise<Response> {
  const first = await send();
  if (first.status !== 401) return first;

  const refreshed = await refresh();
  if (!refreshed) {
    throw new SessionExpiredError();
  }

  const second = await send();
  if (second.status === 401) {
    // Refreshed successfully but the retried call still 401'd (e.g. kicked
    // mid-flight, arch §7 "ถูก kick กลางการใช้งาน") — do not loop again.
    clearAccessToken();
    throw new SessionExpiredError();
  }
  return second;
}

// ---- Public endpoints (no access token needed) ----

export async function signup(req: SignupRequest): Promise<SignupResponse> {
  const res = await rawRequest("/signup", { method: "POST", body: JSON.stringify(req) }, {});
  if (res.status === 201) return (await res.json()) as SignupResponse;
  throw new ApiError(res.status, await parseErrorBody(res), parseRetryAfter(res));
}

export async function login(email: string, password: string, deviceId?: string): Promise<TokenResponse> {
  const req: LoginRequest = {
    email,
    password,
    tokenTransport: "cookie", // web ALWAYS declares cookie transport (H-1)
    ...(deviceId ? { deviceId } : {}),
  };
  const res = await rawRequest("/login", { method: "POST", body: JSON.stringify(req) }, {});
  if (res.status === 200) {
    const body = (await res.json()) as TokenResponse;
    setAccessToken(body.accessToken, body.expiresIn);
    return body;
  }
  throw new ApiError(res.status, await parseErrorBody(res), parseRetryAfter(res));
}

// ---- Authenticated endpoints (Bearer + silent-refresh-then-retry-once) ----

export async function getSessions(): Promise<SessionsResponse> {
  const res = await requestWithRefresh(() =>
    rawRequest("/sessions", { method: "GET" }, { auth: true }),
  );
  if (res.status === 200) return (await res.json()) as SessionsResponse;
  throw new ApiError(res.status, await parseErrorBody(res));
}

export async function logoutDevice(familyId?: string): Promise<void> {
  const res = await rawRequest(
    "/logout",
    { method: "POST", body: JSON.stringify(familyId ? { familyId } : {}) },
    { csrf: true },
  );
  if (res.status !== 204) {
    throw new ApiError(res.status, await parseErrorBody(res));
  }
  clearAccessToken();
}

export async function logoutAll(): Promise<void> {
  const res = await requestWithRefresh(() =>
    rawRequest("/logout-all", { method: "POST" }, { auth: true }),
  );
  if (res.status !== 204) {
    throw new ApiError(res.status, await parseErrorBody(res));
  }
  clearAccessToken();
}

export async function changePassword(req: ChangePasswordRequest): Promise<OkResponse> {
  const res = await requestWithRefresh(() =>
    rawRequest(
      "/change-password",
      { method: "POST", body: JSON.stringify(req) },
      { auth: true, csrf: true },
    ),
  );
  if (res.status === 200) return (await res.json()) as OkResponse;
  throw new ApiError(res.status, await parseErrorBody(res), parseRetryAfter(res));
}
